packages/ckeditor5-select-all/src/selectallcommand.js

```diff
@@ -30,10 +31,20 @@ export default class SelectAllCommand extends Command {
 	 */
 	execute() {
 		const model = this.editor.model;
-		const limitElement = model.schema.getLimitElement( model.document.selection );
+		const selection = model.document.selection;
+		let limitElement = model.schema.getLimitElement( selection );
+		let place = 'in';
+		// If entire element was already selected, try selecting all in a parent limit element (if any).
+		if ( selection.containsEntireContent( limitElement ) && limitElement.root !== limitElement ) {
+			do {
+				// Eventually, the $root (limitElement.root === limitElement) will be a limit.
+				limitElement = limitElement.parent;
+			} while ( !model.schema.isLimit( limitElement ) );
+			place = 'on';
+		}
 
 		model.change( writer => {
-			writer.setSelection( limitElement, 'in' );
+			writer.setSelection( limitElement, place );
 		} );
 	}
```

Comment:
I would propose sth like this:

```javascript
	execute() {
		const model = this.editor.model;
		const selection = model.document.selection;

		model.change( writer => {
			writer.setSelection( getClosestSelection( model, selection ) );
		} );
	}
}

function getClosestSelection( model, selection ) {
	const limitElement = model.schema.getLimitElement( selection );

	// If the entire content of the closest limit element is already selected, try selecting closest object.
	if ( selection.containsEntireContent( limitElement ) ) {
		const objectElement = getObjectElement( model.schema, limitElement );

		if ( objectElement ) {
			return model.createSelection( objectElement, 'on' );
		}
	}

	return model.createSelection( limitElement, 'in' );
}

function getObjectElement( schema, element ) {
	do {
		if ( element.parent ) {
			element = element.parent;
		} else {
			return null;
		}
	} while ( !schema.isObject( element ) );

	return element;
}
```

Reply:
Note:

*   We can use `selection` object, no need for `place` variable;
*   It looks for closest `isObject` not `isLimit`;
*   `getObjectElement` is extracted to helper function to make code more readable;
*   We can check if element is a `root` if it has no `parent`.

Reply:
During a F2F talk with @niegowski we noticed some details that should help implement things even better:

*   I agree that this `do {} while()` loop and `place` variable tandem is hard to read, so let's split things a little bit.
*   Instead of `model.createSelection()`, let's use `writer.createRange()`. Unlike selections, ranges are lightweight.
*   `Schema#isLimit` is a tricky concept because some model elements like `tableRow` are also limits and, for sure, the select all feature should not select all cells in a row. I think (but didn't debug that) the proposed implementation works by accident because the loop starts from `limitElement.parent`. This may backfire in the future if we have some deeper structure of limit elements.  
      
    I think what we need is something smarter than just a limit check. It could be we need to find the closest ancestor that allows text/paragraph, so more likely we need to exclude elements that fail this check `schema.checkChild( limitAncestor, '$text' ) || schema.checkChild( limitAncestor, 'paragraph' )`.  
      
    OTOH this still sounds a little weak because in theory there could be limits that allow widgets only, WDYT @Reinmar?

Reply:
Tip: There's `model.createRange*()` too, so you don't have to access the writer to create a range.

Good catch with the limit thing. I didn't realise that before and it's indeed tricky. I can't think of a better check than this:

*   it's a limit element
*   and it can contain a text or paragraph

It's not perfect but it should work in enough cases to be accepted.

Reply:
Created a follow-up #6829.

Reply:
Thanks! Reading the current docs, I understood that "limit" is exactly the "closest editable ancestor", like "limit of editing scope", which led to this implementation. I naively thought that "widget" could be editable and have something meaningful to select-all as well (because why not).